File: day02/src/main.rs
use std::{
    fs::File,
    io::{prelude::*, BufReader},
    path::Path,
};

fn lines_from_file(filename: impl AsRef<Path>) -> Vec<String> {
    let file = File::open(filename).expect("no such file");
    let buf = BufReader::new(file);
    buf.lines()
        .map(|l| l.expect("Could not parse line"))
        .collect()
}

fn main() {
    let lines = lines_from_file("input.txt");
    let mut count = 0;
    for line in &lines {
        let mut v = line.split(" ");
        let mut limits = v.next().unwrap().split("-");
        let min = limits.next().unwrap().parse().unwrap();
        let max = limits.next().unwrap().parse().unwrap();
        let letter = v.next().unwrap().chars().next().unwrap();
        let password = String::from(v.next().unwrap());
        let letter_match = password.matches(letter).count() as u32;
        if letter_match >= min && letter_match <= max {
            count += 1;
        } 
    }
    println!("part1: {}", count);
    count = 0;
    for line in &lines {
        let mut v = line.split(" ");
        let mut indexes = v.next().unwrap().split("-");
        let first_index: usize = indexes.next().unwrap().parse().unwrap();
        let second_index: usize = indexes.next().unwrap().parse().unwrap();
        let letter = v.next().unwrap().chars().next().unwrap();
        let password = String::from(v.next().unwrap());
        let iter_password = password.as_bytes();
        if iter_password[first_index - 1] != iter_password[second_index - 1] && 
           (iter_password[first_index - 1] as char == letter || iter_password[second_index - 1] as char == letter){
            count += 1;
        } 
    }
    println!("part2: {}", count);
}

File: day02/src/main.py
from pathlib import Path
file = Path("day02/input.txt")

def part1(input):
    valid = 0
    for i in input:
        letterCount = i["text"].count(i["letter"])
        if letterCount <= i["max"] and letterCount >= i["min"]:
            valid += 1
    
    print("Part1", valid)
        
def part2(input):
    valid = 0
    for i in input:
        index1, index2 = map(lambda n: n - 1, [i["min"], i["max"]])
        if (i["text"][index1] == i["letter"]) ^ (i["text"][index2] == i["letter"]):
            valid+= 1

    print("Part2", valid)


if __name__ == "__main__":
    f = open(file)
    input = []

    for line in f.readlines():
        rule, text = line.split(":")
        min, rest = rule.split("-")
        max, letter = rest.split(" ")
        input.append({
            "min": int(min),
            "max": int(max),
            "letter":letter,
            "text": text.strip()
        })

    part1(input)
    part2(input)
    
